Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/model/CompiledLibrary.kt

```diff
@@ -0,0 +1,281 @@
+import org.hl7.elm.r1.CodeDef
+import org.hl7.elm.r1.CodeSystemDef
+import org.hl7.elm.r1.ConceptDef
+import org.hl7.elm.r1.Element
+import org.hl7.elm.r1.ExpressionDef
+import org.hl7.elm.r1.FunctionDef
+import org.hl7.elm.r1.IncludeDef
+import org.hl7.elm.r1.Library
+import org.hl7.elm.r1.ParameterDef
+import org.hl7.elm.r1.UsingDef
+import org.hl7.elm.r1.ValueSetDef
+import org.hl7.elm.r1.VersionedIdentifier
+
+@Suppress("TooManyFunctions")
+class CompiledLibrary {
+    var identifier: VersionedIdentifier? = null
+    var library: Library? = null
+    private val namespace: MutableMap<String, Element> = HashMap()
+    val operatorMap: OperatorMap = OperatorMap()
+    private val functionDefs: MutableMap<Operator, FunctionDef> = HashMap()
+    private val conversions: MutableList<Conversion> = ArrayList()
+
+    private fun checkNamespace(identifier: String) {
+        val existingResolvedIdentifierContext = resolve(identifier)
+        existingResolvedIdentifierContext.exactMatchElement.ifPresent {
+            throw IllegalArgumentException(
+                String.format(
+                    Locale.US,
+                    "Identifier %s is already in use in this library.",
+                    identifier
+                )
+            )
+        }
+    }
+
+    fun add(using: UsingDef) {
+        checkNamespace(using.localIdentifier)
+        namespace[using.localIdentifier] = using
+    }
+
+    fun add(include: IncludeDef) {
+        checkNamespace(include.localIdentifier)
+        namespace[include.localIdentifier] = include
+    }
+
+    fun add(codesystem: CodeSystemDef) {
+        checkNamespace(codesystem.name)
+        namespace[codesystem.name] = codesystem
+    }
+
+    fun add(valueset: ValueSetDef) {
+        checkNamespace(valueset.name)
+        namespace[valueset.name] = valueset
+    }
+
+    fun add(code: CodeDef) {
+        checkNamespace(code.name)
+        namespace[code.name] = code
+    }
+
+    fun add(concept: ConceptDef) {
+        checkNamespace(concept.name)
+        namespace[concept.name] = concept
+    }
+
+    fun add(parameter: ParameterDef) {
+        checkNamespace(parameter.name)
+        namespace[parameter.name] = parameter
+    }
+
+    fun add(expression: ExpressionDef) {
+        if (expression is FunctionDef) {
+            // Register the operator signature
+            add(expression, Operator(expression))
+        } else {
+            checkNamespace(expression.name)
+            namespace[expression.name] = expression
+        }
+    }
+
+    fun remove(expression: ExpressionDef) {
+        require(expression !is FunctionDef) { "FunctionDef cannot be removed." }
+        namespace.remove(expression.name)
+    }
+
+    private fun ensureLibrary(operator: Operator) {
+        // Functions can be defined in an anonymous library
+        if (this.identifier != null && identifier!!.id != null) {
+            if (operator.libraryName == null) {
+                operator.libraryName = identifier!!.id
```

Comment:
Question:  From what I understand, Kotlin's !! is meant to throw an NPE if the variable in question is null.  Given Kotlin has "?" under which situations do we want to deliberately use "!!" in the CQL project?

Reply:
I'll take a look at this again. I removed most of the cases where this happened in the auto-generated code. Looks like I missed this one.